Documents the operator's support for TLS re-encryption. (#47565)
keycloak/keycloak: docs/guides/operator/basic-deployment.adoc

NOTE: The operator annotates the Ingress to match expectations for TLS passthrough or TLS termination on OpenShift with the default IngressClass. 
See below for more on TLS termination. 

==== TLS Termination with default Ingress

> When using the operator-provided ingress (`spec.ingress.enabled: true`) there are a lot of Keycloak CR options (like service and ingress labels and annotations) which allows you to customize your ingress behaviour. This section show how to achive below TLS Termination when using Openshift:

Reviewer: ```suggestion
When using the operator-provided ingress (`spec.ingress.enabled: true`) there are a lot of Keycloak CR options (like service and ingress labels and annotations) which allow you to customize your ingress behaviour. This section shows how to achieve the following TLS Termination when using Openshift:
```

> When using the operator-provided ingress (`spec.ingress.enabled: true`) there are a lot of Keycloak CR options (like service and ingress labels and annotations) which allow you to customize your ingress behaviour. This section shows how to achieve the following TLS Termination when using Openshift:

Reviewer 2: ```suggestion
When using the operator-provided ingress (`spec.ingress.enabled: true`) there are a lot of Keycloak CR options (like service and ingress labels and annotations) which allow you to customize your ingress behaviour. This section shows how to achieve the following TLS Termination when using OpenShift:
```

- <<passthrough, Passthrough>>
- <<edge_cluster_domain, Edge using cluster domain and certificate>>
- <<edge_custom_domain, Edge using custom domain and certificate>>
- <<reencrypt_cluster_domain, Reencrypt using cluster domain and certificate>>
- <<reencrypt_custom_domain, Reencrypt using custom domain and certificate>>

> For other platforms or IngressControllers, you can use the provided examples as a starting point.

Reviewer: Thank you this looks good.

[[passthrough]]
===== Passthrough

It is enabled when you associate a `tlsSecret` with the `http` configuration and leave Ingress enabled without specifying a `tlsSecret` on it. 

[source,yaml]
----
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  http:
    tlsSecret: example-tls-secret
  ingress:
    enabled: true
  hostname:
    hostname: test.keycloak.org
----

NOTE: On OpenShift you are not allowed to use wildcard certificates on passthrough Routes with HTTP/2 enabled. For wildcard certificates use edge or reencrypt termination.

===== Edge

There are two ways to achieve TLS Edge. In both it is needed to enable HTTP access.

[[edge_cluster_domain]]
*Using cluster domain and certificate*

To use cluster domain do not provide a tls secret for the `ingress` and ensure to use cluster domain as `hostname`. Since no certificate is provided to the `ingress` the cluster certificate is used.

[source,yaml]
----
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  http:
    httpEnabled: true
  ingress:
    enabled: true
  hostname:
    hostname: keycloak.<cluster-domain> <1>
  proxy:

>     headers: xforwarded # double check your reverse proxy sets and overwrites the X-Forwarded-* headers

Reviewer 2: I consider "double check your reverse proxy" a bit too vague for a documentation.
What should people do here?

Reviewer 2: Happy to handle this in a follow-up issue.

Author: > I consider "double check your reverse proxy" a bit too vague for a documentation.
What should people do here?

Perhaps the text bellow is better:

Check your load balancer or reverse proxy configuration to determine whether it utilizes the Forwarded (RFC 7239) or X-Forwarded-* (e.g., X-Forwarded-For) mechanism for header propagation, so set this option accordingly

Author: But a bit large for a comment in a code block

There are two ways to achieve TLS Reencryption. In both it is needed to annotate the `ingress` to specify the desired termination type since operator still lacks logic to set itself the reencrypt termination.

[[reencrypt_cluster_domain]]
*Using cluster domain and certificate*

To use cluster domain do not provide a tls secret for the `ingress` and ensure to use cluster domain as `hostname`. Since no certificate is provided to the `ingress` the cluster certificate is used.

[source,yaml]
----
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  http:
    annotations:
      service.alpha.openshift.io/serving-cert-secret-name: svc-tls-secret <1>
      tlsSecret: svc-tls-secret <2>
  ingress:
    annotations:
      route.openshift.io/termination: reencrypt <3>
    enabled: true
  hostname:
    hostname: keycloak.<cluster-domain> <4>
  proxy:
    headers: xforwarded # double check your reverse proxy sets and overwrites the X-Forwarded-* headers
----
<1> This will add the annotation to the service generated by the operator, so cluster will create a tls secret named svc-tls-secret
<2> name of the tls secret created by the cluster, as defined in <1>

> <3> required because the operator only has built-in openshift logic for passthrough and edge. 

Reviewer 2: ```suggestion
<3> required because the operator only has built-in OpenShift logic for passthrough and edge. 
```

<4> Must use the cluster domain

[[reencrypt_custom_domain]]
*Using custom domain and certificate*

To use custom domain provide distinct tls secrets for the `ingress` and `http` and ensure that `hostname` matches the `ingress` certificate.

[source,yaml]
----
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  http:
    annotations:
      service.alpha.openshift.io/serving-cert-secret-name: svc-tls-secret <1>
      tlsSecret: svc-tls-secret <2>
  ingress:
    annotations:
      route.openshift.io/termination: reencrypt <3>
    tlsSecret: ingress-tls-secret <4>
    enabled: true
  hostname:
    hostname: test.keycloak.org <5>
  proxy:
    headers: xforwarded # double check your reverse proxy sets and overwrites the X-Forwarded-* headers
----

<1> this will add the annotation to the service generated by the operator, so cluster will create a tls secret named svc-tls-secret
<2> name of the tls secret created by the cluster, as defined in <1>

> <3> required because the operator only has built-in openshift logic for passthrough and edge.

Reviewer 2: ```suggestion
<3> required because the operator only has built-in OpenShift logic for passthrough and edge.
```
